### PR TITLE
add dnn graph

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?
1. add dnn graph
2. add lenet5, vgg16, resnet50 dnn graph implementation

## How was this patch tested?

unit test

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

**Comment by zhangxiaoli73** [comment]
dnn graph just extends static graph

**Comment by zhangxiaoli73** [comment]
use mkldnn Perf script to test graph train performance, results as following:

resnet50 mkldnn graph:
average throughput:  9.5374  iters:  50
resnet50 mkldnn:
average throughput:  11.2544  iters:  50

vgg16 mkldnn graph:
average throughput:  3.9068  iters:  50
vgg16 mkldnn:
average throughput:  4.1572  iters:  50

**Comment by wzhongyuan** [comment]
@zhangxiaoli73   Why DNN graph throughtput is less than non-graph ?

**Comment by zhangxiaoli73** [comment]
@wzhongyuan. Wrong to add fuse non-graph performance data. Actually, there should be no fuse when training. Have updated.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+    other: Activity, otherFormats: Array[MemoryData]): Activity = {
+    val realOthers = if (otherFormats.length > 0) {
+      reorderManager.infer(otherFormats, realFormats, other)
+    } else {
+      other
+    }
+    super.accActivity(activity, realOthers)
+  }
+
+  private def compile() : Unit = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: is this method thread safe ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+
+  if (enableExcludeChecking) {
+    excludeInvalidLayers(forwardExecution.map {_.element})
+  }
+
+  buildBackwardGraph()
+
+  override def updateOutput(input: Activity): Activity = {
+    // compile dnn graph according to model status
+    if (!compiled) compile()
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: If there're a Graph `graph` which contains another Graph named `subGraph`. So when the `subGraph.compiled` is set?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala (hunk 1 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: yeah, I think so.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala (hunk 2 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: it should be very intuitive one graph contains another, right ?

Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: Yeah, in long term, we need to support this case. Maybe I can break down sub dnnGraph to get its inputs/outputs nodes and merge those nodes to dnnGraph when building dnnGraph.
Now, dnn layers are limited and no such case that is much necessary to do this change (maybe over design ).
So I doubt if it is worth doing now. Maybe we can add it to "todo list".

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/lenet/LeNet5.scala [hunk 3]
```diff
@@ -103,4 +104,22 @@ object LeNet5 {
       .add(mkldnn.ReorderMemory(mkldnn.HeapData(outputShape, Memory.Format.nc)))
     model
   }
+
+  def dnnGraph(batchSize: Int, classNum: Int): mkldnn.DnnGraph = {
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/lenet/LeNet5.scala]: How about add an option to example?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/models/Vgg16.scala [hunk 4]
```diff
@@ -90,4 +90,65 @@ object Vgg_16 {
       conv
     }
   }
+
+  def graph(batchSize: Int, classNum: Int, hasDropout: Boolean = true): DnnGraph = {
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/models/Vgg16.scala]: How about add an option to example?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Perf.scala [hunk 5]
```diff
@@ -256,6 +266,140 @@ object ResNet {
     model
   }
 
+  def graph(batchSize: Int, classNum: Int, opt: Table): DnnGraph = {
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Perf.scala]: How about add an option to example?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+  }
+
+  // change nn identity to mkldnn identity
+  private def toDnnIdentity(model: nn.Identity[Float])
+  : AbstractModule[Activity, Activity, Float] = {
+    mkldnn.Identity[Float]().setName(model.getName())
+      .asInstanceOf[AbstractModule[Activity, Activity, Float]]
+  }
+
+  // if node has no previous node, then it will just use input as real input
```
[the remark this answers is not in the document]
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: good idea.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+          case Some(i) =>
+            if (i == 1 && curNode.element.output.isTensor) {
+              curGradOutput = addActivity(curGradOutput, realGradOutputFormats,
+                otherActivity, format)
+            } else {
+              if (curNode.element.output.isTable && curGradOutput == null) {
+                curGradOutput = T()
+              }
+              val curActivity = curGradOutput.toTable.getOrElse[Activity](i, null)
+              curGradOutput.toTable(i) = addActivity(curActivity, realGradOutputFormats,
```
[the remark this answers is not in the document]
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: yeah, good idea

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+        }
+      })
+
+    if (curNode.element.output.isTable) {
+      addZeroTensorToMissingGradOutput(curNode.element.output.toTable, curGradOutput.toTable)
+    }
+    curGradOutput
+  }
+
+  private def addActivity(activity: Activity, realFormats: Array[MemoryData],
```
[the remark this answers is not in the document]
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: no, get the right format and then add those two activities. (call super.accActivity)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+
+  // change nn identity to mkldnn identity
+  private def toDnnIdentity(model: nn.Identity[Float])
+  : AbstractModule[Activity, Activity, Float] = {
+    mkldnn.Identity[Float]().setName(model.getName())
+      .asInstanceOf[AbstractModule[Activity, Activity, Float]]
+  }
+
+  // if node has no previous node, then it will just use input as real input
+  private def findDnnInput(node: ModuleNode[Float], input: Activity): Activity = {
```
[the remark this answers is not in the document]
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: I think it's more clear. Distinct from "findInput" in Graph

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+      } else {
+        (T.seq(prevActivitiesAndFormats.map(m => m._1)),
+          prevActivitiesAndFormats.map(m => m._2).toArray.flatMap(_.toSeq))
+      }
+      reorderManager.infer(inputAndFormat._2, realInputFormats, inputAndFormat._1)
+    }
+    nodeInput
+  }
+
+  private def findDnnGradOutput(curNode: ModuleNode[Float], gradOutput: Activity,
```
[the remark this answers is not in the document]
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: same

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+  private def findInputFormats(node: ModuleNode[Float], inputs: Array[MemoryData])
+    : Array[MemoryData] = {
+    if (node.prevNodes.isEmpty) {
+      inputs
+    } else {
+      val prevFormats = node.prevNodesAndEdges
+        .filterNot(n => n._1.element.isInstanceOf[ControlDependency[Float]])
+        .map(n => {
+          // output is tensor and fromIndex number is 1
+          if (n._1.element.asInstanceOf[MklDnnModule].outputFormats().length == 1
```
[the remark this answers is not in the document]
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: yeah, good idea

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/lenet/LeNet5.scala (hunk 3 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/models/lenet/LeNet5.scala]: ok. How about using "graph" & "mkldnn" engine type to call dnnGraph?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/models/Vgg16.scala (hunk 4 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/models/Vgg16.scala]: ok

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Perf.scala (hunk 5 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Perf.scala]: ok

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/lenet/Train.scala [hunk 12]
```diff
@@ -49,7 +49,10 @@ object Train {
         Module.load[Float](param.modelSnapshot.get)
       } else {
         if (param.graphModel) {
-          LeNet5.graph(classNum = 10)
+          Engine.getEngineType() match {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/lenet/Train.scala]: This is hard to understand for user.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/TrainImageNet.scala [hunk 13]
```diff
@@ -90,8 +90,13 @@ object TrainImageNet {
 
             curModel
           case MklDnn =>
-            nn.mkldnn.ResNet(param.batchSize / Engine.nodeNumber(), param.classes,
-              T("depth" -> 50, "dataSet" -> ImageNet))
+            if (param.graphModel) {
+              nn.mkldnn.ResNet.graph(param.batchSize / Engine.nodeNumber(), param.classes,
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/TrainImageNet.scala]: I think this is hard to understand for user

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/lenet/LeNet5.scala (hunk 3 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/models/lenet/LeNet5.scala]: yeah, good idea. Changed

**Pull request merged**